### PR TITLE
Stop only the pebble services defined in the container

### PR DESCRIPTION
**Pull request by michaeldmitry**
## Issue
In case of using a worker workload image that doesn't define a pebble service (e.g. the upstream docker image), the shared worker raises an error as it tries to stop the pebble services defined in the layer generated by the charm, which are not necessarily running in the container.

## Solution
Refactor the `stop()` method to only look at the services defined in the container, as we only care about stopping the services that are currently running.

### CODE REVIEW

File: src/coordinated_workers/worker.py [hunk 1]
```diff
@@ -546,19 +546,20 @@ def stop(self):
 
         Assumes that pebble can connect.
         """
+        curr_services = self._container.get_plan().services
```
Comment (mmkay) [inline src/coordinated_workers/worker.py]: I think this would make more sense to call it `container_services`. That way it'll be clearer which services we take into account here.

File: src/coordinated_workers/worker.py [hunk 2]
```diff
@@ -546,19 +546,20 @@ def stop(self):
...
-            else self._container.get_plan().services
+            else curr_services
         )
 
         if not services:
             logger.warning("nothing to stop: no services found in layer or plan")
             return
-
-        self._container.stop(*services)
+        if curr_services:
```
Comment (mmkay) [inline src/coordinated_workers/worker.py]: Let's prepend it with a comment explaining what happens here. Perhaps:

```
# if the pebble layer is defined in the charm, but the container isn't a rock, it will not have any pebble layer. Stopping services would then throw an exception.
```

File: src/coordinated_workers/worker.py (hunk 1 above)
Comment (mmkay) [inline src/coordinated_workers/worker.py]: Perhaps we can also exit early? If there's no services defined in the container we'll have nothing to stop anyway.

File: src/coordinated_workers/worker.py [hunk 3]
```diff
@@ -546,19 +546,20 @@ def stop(self):
 
         Assumes that pebble can connect.
         """
+        curr_services = self._container.get_plan().services
         # we might be unable to retrieve self.pebble_layer if something goes wrong generating it
         # for example because we're being torn down and the environment is being weird
         services = tuple(
             self.pebble_layer.services
```
Comment (mmkay) [inline src/coordinated_workers/worker.py]: I wonder if we really need to look at the charm pebble layer at all. We actually end up trying to stop any services that are defined in the pebble layer, but only if there are any services already defined in the container.

**Review by mmkay — COMMENTED** [review]
Left a few comments that mostly aim at making this piece more readable. Because there are two pebble layers in question, one in the charm and another one in the container, this can get tricky to understand.

Maybe we could have a test case for this edge case too?

File: src/coordinated_workers/worker.py [hunk 4]
```diff
@@ -546,18 +546,10 @@ def stop(self):
...
         """
-        # we might be unable to retrieve self.pebble_layer if something goes wrong generating it
-        # for example because we're being torn down and the environment is being weird
-        services = tuple(
-            self.pebble_layer.services
-            if self.pebble_layer
-            else self._container.get_plan().services
-        )
-
+        services = tuple(self._container.get_plan().services)
```
Comment (mmkay) [inline src/coordinated_workers/worker.py:549]: Minor: do we still need the `tuple(`?

**Review by mmkay — APPROVED** [review]
